sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala

```diff
@@ -45,7 +46,12 @@ class SQLBuilder(logicalPlan: LogicalPlan) extends Logging {
   def this(df: Dataset[_]) = this(df.queryExecution.analyzed)
 
   private val nextSubqueryId = new AtomicLong(0)
+  private val nextGenAttrId = new AtomicLong(0)
   private def newSubqueryName(): String = s"gen_subquery_${nextSubqueryId.getAndIncrement()}"
+  private val exprIdMap: Map[Long, Long] = Map.empty[Long, Long]
+  private def normalizedName(n: NamedExpression): String = synchronized {
+    "gen_attr_" + exprIdMap.getOrElseUpdate(n.exprId.id, nextGenAttrId.getAndIncrement())
+  }
```

Comment:
Now, the generated attributes also have stable identifiers like generated subqueries.